Truncate the starts of file paths instead of the ends in picker

Omnikar opened the pull request:
Fixes #945.

pickfire commented on helix-tui/src/buffer.rs:
```diff
@@ -289,28 +292,60 @@ impl Buffer {
-                self.content[i].reset();
+        } else {
+            let mut truncated = false;
+            let mut end = Vec::new();
```
I think we can do better than collecting iterating over this twice and having a separate allocation.

I think probably can just get the index and display the rest. Didn't look closely.

Omnikar replied:
Since individual graphemes can have different widths, I'm not sure how to do this without iterating over it in reverse and checking the total width.

> single character for each directory in path

Ah, that's a good idea.

archseer replied:
You should be able to just calculate the total width using `unicode_width::UnicodeWidthStr`, no need to segment it into graphemes: https://github.com/helix-editor/helix/blob/e505bf2b4809681840d1684a23ad21255fca290d/helix-core/src/graphemes.rs#L30

We re-export the library as `helix_core::unicode::width`

archseer replied:
Ah yeah, it's the same method you're already using (`.width()`). Should be enough to call  `string.width()`, then if it's larger than `width` we iterate graphemes from reverse, rendering them into `content` (without using the `end` vector)

archseer replied:
I pushed a rewrite with ^ but it's untested. Probably needs more work.

Omnikar replied:
Yeah… this probably isn't what's supposed to happen 😄 
<img width="364" alt="image" src="https://user-images.githubusercontent.com/79615262/140236895-ab9864b1-7973-4484-83ee-46e3e38358fa.png">

Omnikar replied:
And it panics when the paths are long enough to actually need to be truncated.

Omnikar replied:
Ah, addressing [this](https://github.com/helix-editor/helix/pull/951#discussion_r741586853) fixes the panic.

archseer replied:
Ah, I guess that codepath always gets called even if the string is shorter.

archseer replied:
We'd preferably use the original code under `!truncate_start` if the width is shorter

Omnikar replied:
I think I've figured out what to do.

archseer commented on helix-tui/src/buffer.rs:
```diff
@@ -317,34 +317,29 @@ impl Buffer {
-                self.content[index].set_symbol(s);
-                self.content[index].set_style(style);
-                for i in index + 1..index + width {
+                let start = index - width;
```
We need to stop here if `start < start_index`

Omnikar commented:
Wait a minute, it stopped working before I pushed it? I tested it and it worked before I deleted `let mut x_offset = max_offset;` because it was no longer being used. But after I deleted it, it started subtract overflowing?

Omnikar commented:
Ohhhh, it was because `x_offset` shadowed another `x_offset` variable… which means that the compiler was wrong in warning me that the new `x_offset` is unused? 🤔

Omnikar commented:
Alright, now it should work.

Omnikar commented:
OK, *now* it should work.
